src/coercions.rs

```diff
@@ -0,0 +1,219 @@
+    let mut root = HashMap::new();
+
+    for field in schema.get_fields() {
+        append_coercion(&mut root, field)
+    }
+
+    CoercionTree { root }
+}
+
+fn append_coercion(context: &mut HashMap<String, CoercionNode>, field: &DeltaField) {
+    match field.get_type() {
+        DeltaDataType::primitive(primitive_type) if primitive_type == "string" => {
+            context.insert(
+                field.get_name().to_string(),
+                CoercionNode::Coercion(Coercion::ToString),
+            );
+        }
+        DeltaDataType::primitive(primitive_type) if primitive_type == "timestamp" => {
+            context.insert(
+                field.get_name().to_string(),
+                CoercionNode::Coercion(Coercion::ToTimestamp),
+            );
+        }
+        DeltaDataType::r#struct(schema) => {
+            let mut nested_context = HashMap::new();
+            for nested_field in schema.get_fields() {
+                append_coercion(&mut nested_context, nested_field);
+            }
+            if !nested_context.is_empty() {
+                let tree = CoercionTree {
+                    root: nested_context,
+                };
+                context.insert(field.get_name().to_string(), CoercionNode::Tree(tree));
+            }
+        }
+        _ => {
+            // noop for now
+            // add more data type coercions as necessary
+        }
+    }
+}
+
+/// Applies all data coercions specified by the [`CoercionTree`] to the [`Value`].
+/// Though it does not currently, this function should approximate or improve on the coercions applied by [Spark's `from_json`](https://spark.apache.org/docs/latest/api/sql/index.html#from_json)
+pub(crate) fn coerce(value: &mut Value, coercion_tree: &CoercionTree) {
+    if let Some(context) = value.as_object_mut() {
+        for (field_name, coercion) in coercion_tree.root.iter() {
+            apply_coercion(context, field_name, coercion);
+        }
+    }
+}
+
+fn apply_coercion(context: &mut Map<String, Value>, field_name: &str, node: &CoercionNode) {
+    let opt = context.get_mut(field_name);
+
+    if let Some(value) = opt {
+        match node {
+            CoercionNode::Coercion(Coercion::ToString) => {
+                let replacement = if value.is_string() {
+                    None
+                } else {
+                    Some(value.to_string())
+                };
+
+                if let Some(coerced) = replacement {
+                    context.insert(field_name.to_string(), Value::String(coerced));
+                }
+            }
+            CoercionNode::Coercion(Coercion::ToTimestamp) => {
+                let replacement: Option<i64> = if let Some(s) = value.as_str() {
+                    // The delta timestamp data type must be set as microseconds since epoch.
+                    // If we have a string, try to convert it.
+                    // If conversion fails, leave it alone. It'll come out null in delta.
```

Comment:
Or maybe let's move to dead letters queue? How's that different from invalid data format, e.g. expected string, passed integer

Reply:
The difference with coercions is that Spark does these already, and leaves uncoercable fields null rather than not inserting the row at all. I feel like we have a closer match to that behavior by letting the values go null as they do in Spark. We actuallly have cases today where clients send strings like `"123"` for columns that are typed as "integer". Spark does not coerce these, and the value is row is stored with a null value for the int column. If we were to dead letter these instead, it might be a disruptive change.

Reply:
I see, if spark works this way then I'm fine with it too